smarts/env/gymnasium/wrappers/metric/costs.py

```diff
@@ -544,33 +544,44 @@ def get_dist(road_map: RoadMap, point_a: Point, point_b: Point) -> float:
         float: Shortest road distance between two points in the road map.
     """
 
-    def _get_dist(start: Point, end: Point) -> float:
-        mission = Mission(
-            start=Start(
-                position=start.as_np_array,
-                heading=Heading(0),
-                from_front_bumper=False,
-            ),
-            goal=PositionalGoal(
-                position=end,
-                radius=3,
-            ),
-        )
-        plan = Plan(road_map=road_map, mission=mission, find_route=False)
-        plan.create_route(mission=mission, radius=20)
-        from_route_point = RoadMap.Route.RoutePoint(pt=start)
-        to_route_point = RoadMap.Route.RoutePoint(pt=end)
-
-        dist_tot = plan.route.distance_between(
-            start=from_route_point, end=to_route_point
-        )
-        if dist_tot == None:
-            raise CostError("Unable to find road on route near given points.")
-        elif dist_tot < 0:
-            raise CostError(
-                "Path from start point to end point flows in "
-                "the opposite direction of the generated route."
-            )
-        return dist_tot
-
-    return _get_dist(point_a, point_b)
+    mission = Mission(
+        start=Start(
+            position=point_a.as_np_array,
+            heading=Heading(0),
+            from_front_bumper=False,
+        ),
+        goal=PositionalGoal(
+            position=point_b,
+            radius=2,
+        ),
+    )
+    plan = Plan(road_map=road_map, mission=mission, find_route=False)
+    plan.create_route(mission=mission, radius=20)
+    assert isinstance(plan.route, RoadMap.Route)
+    from_route_point = RoadMap.Route.RoutePoint(pt=point_a)
+    to_route_point = RoadMap.Route.RoutePoint(pt=point_b)
+
+    dist_tot = plan.route.distance_between(
+        start=from_route_point, end=to_route_point
+    )
+    if dist_tot == None:
+        raise CostError("Unable to find road on route near given points.")
+    elif dist_tot < 0:
+        # This happens when agent overshoots the goal position while 
+        # remaining outside the goal capture radius at all times. Default 
+        # positional goal radius is 2m.
+        dist_tot = abs(dist_tot)
+
+    # Account for agent ending in a different lane but in the same road as
+    # the goal position. 
+    start_lane = plan.route.start_lane
+    end_lane = plan.route.end_lane
+    lane_error = abs(start_lane.index-end_lane.index)
+    if len(plan.route.roads) == 1 and lane_error > 0:
```

Comment:
Does this need to support routes with more than 1 road?

Reply:
Here we intend to add lane error penalty only when the ego is in the same road as the goal position, but in a different lane. Hence, we check for `len(plan.route.roads) == 1` where the computed `route` is between the final ego position and the goal position. If there are more than one road in the `route`, the lane error penalty becomes ambiguous and thus it is not implemented.